main.c

```diff
@@ -677,123 +718,157 @@ node_compare (const void *n1, const void *n2)
   return strcmp (node1->name, node2->name) == 0 ? true : false;
 }
 
+static void
+cleanup_node_initp (struct ovl_node **p)
+{
+  struct ovl_node *n = *p;
+  if (n == NULL)
+    return;
+  if (n->children)
+    hash_free (n->children);
+  free (n->name);
+  free (n->path);
+  free (n);
+}
+
+#define cleanup_node_init __attribute__((cleanup (cleanup_node_initp)))
+
 static struct ovl_node *
 make_whiteout_node (const char *path, const char *name)
 {
-  struct ovl_node *ret = calloc (1, sizeof (*ret));
+  struct ovl_node *ret_xchg;
+  cleanup_node_init struct ovl_node *ret = NULL;
+
+  ret = calloc (1, sizeof (*ret));
   if (ret == NULL)
-    {
-      errno = ENOMEM;
-      return NULL;
-    }
+    return NULL;
+
   ret->name = strdup (name);
   if (ret->name == NULL)
-    {
-      free (ret);
-      errno = ENOMEM;
       return NULL;
-    }
+
   ret->path = strdup (path);
   if (ret->path == NULL)
+    return NULL;
+
+  ret->whiteout = 1;
+
+  ret_xchg = ret;
+  ret = NULL;
+
+  return ret_xchg;
+}
+
+static ssize_t
+safe_read_xattr (char **ret, int sfd, const char *name, size_t initial_size)
+{
+  cleanup_free char *buffer = NULL;
+  size_t current_size;
+  ssize_t s;
+
+  current_size = initial_size;
+  buffer = malloc (current_size + 1);
+  if (buffer == NULL)
+    return -1;
+
+  while (1)
     {
-      free (ret->name);
-      free (ret);
-      errno = ENOMEM;
-      return NULL;
+      char *tmp;
+
+      s = fgetxattr (sfd, name, buffer, current_size);
+      if (s < 0)
+        break;
+      if (s < current_size)
+        break;
+
+      current_size *= 2;
+      tmp = realloc (buffer, current_size + 1);
+      if (tmp == NULL)
```

Comment:
Isn't this a potential leak of buffer here, on realloc failure.

Reply:
`buffer` is marked with `cleanup_free`, so it is automatically freed when we leave the scope.

In facts, to avoid it is freed on success, we need to set it to NULL